gobblin-yarn/src/main/java/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -288,6 +331,38 @@ public void handleApplicationReportArrivalEvent(ApplicationReportArrivalEvent ap
     }
   }
 
+  private Optional<ApplicationId> getApplicationId() throws YarnException, IOException {
+    Optional<ApplicationId> reconnectableApplicationId = getReconnectableApplicationId();
+    if (reconnectableApplicationId.isPresent()) {
+      LOGGER.info("Found reconnectable application with application ID: " + reconnectableApplicationId.get());
+      return reconnectableApplicationId;
+    }
+
+    LOGGER.info("No reconnectable application found so submitting a new application");
+    return Optional.of(setupAndSubmitApplication());
+  }
+
+  private Optional<ApplicationId> getReconnectableApplicationId() throws YarnException, IOException {
+    Optional<String> existingApplicationId = this.applicationIdStore.get();
+    if (!existingApplicationId.isPresent()) {
+      return Optional.absent();
+    }
+
+    List<ApplicationReport> applicationReports =
```

Comment:
The Yarn API does not support creating an `ApplicationId` object from a string ID, while `getApplicationReport` needs an `ApplicationId` object.